core/src/Kokkos_CudaSpace.hpp

```diff
@@ -473,20 +473,10 @@ struct DeepCopy<MemSpace1, MemSpace2, ExecutionSpace,
     DeepCopyCuda(dst, src, n);
   }
 
-  inline DeepCopy(const ExecutionSpace& exec, void* dst, const void* src,
+  inline DeepCopy(const ExecutionSpace& /*exec*/, void* dst, const void* src,
                   size_t n) {
-    exec.fence(fence_string());
     DeepCopyAsyncCuda(dst, src, n);
```

Comment:
This is a bit confusing, but should this actually call `DeepCopyCuda(dsat, src, n)` since that is now truly asynchronous? Or is it intentionally being left as is, because there's no way for the caller to synchronize on this call other than a full `Kokkos::fence()`?

More broadly, is it truly considered legitimate Kokkos usage to make a deep copy call between (e.g.) CUDA memory spaces with some arbitrary non-CUDA execution space? If it were specifically `Serial`, I would support allowing that, since its 'instance' fence could devolve to a global fence that would cover the device streams, but I'm not sure that generalizes to a hypothetical future where we allow mixed builds of say `OpenMP` on the CPU and `CUDA` on the GPU.

Reply:
Just for clarification, this is the case where the execution space is not `Kokkos::Cuda` so we can't run the `memcpy` on the provided execution space.

The fence before `DeepCopyAsyncCuda` is debatable and there might be a good reason for it. In my opinion, we expect for the three-argument `deep_copy`/`DeepCopy` with execution space that the operation is complete when we fence the given execution space. Since the actual `memcpy` is not executed using that execution space, we have to fence the copy stream implicitly. Thus, we can't use `DeepCopyCuda(dst, src, n)` here.
The default `ExecutionSpace` argument is given by the execution space associated with the memory space of the destination which makes sense to me since you expect to fence that space to make sure the `memcpy` has been completed. Of course, we never want to really call this overload since following this philosophy requires it to fence more (and not be asynchronous at all). I think that arguing about this behavior (and finding a better default) belongs to another issue, though.

Reply:
We promise to fence the incoming execution space in this case, see https://github.com/kokkos/kokkos/wiki/Kokkos::deep_copy, so I reverted these changes.